scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala

```diff
@@ -495,6 +500,103 @@ object TableRowJsonIO {
   )
 }
 
+final case class BigQueryDynamicTable[T: Coder](writer: beam.BigQueryIO.Write[T])(
+  tableFn: ValueInSingleWindow[T] => TableDestination
+) extends ScioIO[T] {
+
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+  override type ReadP = Nothing // WriteOnly
+  override type WriteP = BigQueryDynamicTable.WriteParam
+
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+    throw new UnsupportedOperationException("BigQueryDynamicTable is read-only")
+
+  override protected def write(self: SCollection[T], params: WriteP): Tap[Nothing] = {
+    if (self.context.isTest) {
+      throw new NotImplementedError(
+        "BigQuery with dynamic table destinations cannot be used in a test context"
+      )
+    } else {
+      val destinations = DynamicDestinationsUtil.tableFn(tableFn, params.schema)
+
+      var transform = writer.to(destinations)
+
+      if (params.createDisposition != null) {
+        transform = transform.withCreateDisposition(params.createDisposition)
+      }
+      if (params.writeDisposition != null) {
+        transform = transform.withWriteDisposition(params.writeDisposition)
+      }
+
+      transform = params.extendedErrorInfo match {
+        case Disabled => transform
+        case Enabled  => transform.withExtendedErrorInfo()
+      }
+
+      val wr = self.applyInternal(transform)
+
+      params.insertErrorTransform(params.extendedErrorInfo.coll(self.context, wr))
+    }
+    EmptyTap
+  }
+  override def tap(params: Nothing): Tap[Nothing] = EmptyTap
+}
+
+object BigQueryDynamicTable {
+  trait WriteParam {
+    val schema: TableSchema
+    val writeDisposition: WriteDisposition
+    val createDisposition: CreateDisposition
+    val extendedErrorInfo: ExtendedErrorInfo
+    val insertErrorTransform: SCollection[extendedErrorInfo.Info] => Unit
+  }
+
+  object WriteParam extends Writes.WriteParamDefauls {
+    @inline final def apply(
+      schema: TableSchema,
+      wd: WriteDisposition,
+      cd: CreateDisposition,
+      ei: ExtendedErrorInfo
+    )(it: SCollection[ei.Info] => Unit): WriteParam = new WriteParam {
+      val schema: TableSchema = schema
```

Comment:
this seems odd; ref itself?

Reply:
this follows the same pattern as the other WriteParam constructors in BigQueryIO. See below:

https://github.com/spotify/scio/blob/master/scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala#L221:L237